crates/sim2h/src/connection_mgr.rs

```diff
@@ -0,0 +1,319 @@
+                    error!("socket read error {} {:?}", uri, e);
+                    let _ = evt_send.send(ConMgrEvent::Disconnect(uri.clone(), Some(e.into())));
+                    // end task
+                    return;
+                }
+            }
+        }
+
+        // if we did work we might have more work to do,
+        // if not, let this task get parked for a time
+        if did_work {
+            tokio::task::yield_now().await;
+        } else {
+            tokio::time::delay_for(std::time::Duration::from_millis(5)).await;
+        }
+    }
+}
+
+/// internal actually spawn the above wss_task into the tokio runtime
+fn spawn_wss_task(uri: Lib3hUri, wss: TcpWss, evt_send: EvtSend) -> CmdSend {
+    let (cmd_send, cmd_recv) = tokio::sync::mpsc::unbounded_channel();
+    tokio::task::spawn(wss_task(uri, wss, evt_send, cmd_recv));
+    cmd_send
+}
+
+/// internal result enum for connection mgr processing loop
+enum ConMgrResult {
+    DidWork,
+    NoWork,
+    EndTask,
+}
+
+use ConMgrResult::*;
+
+/// internal loop for processing the connection mgr
+async fn con_mgr_task(mut con_mgr: ConnectionMgr, weak_ref_dummy: Weak<()>) {
+    loop {
+        if let None = weak_ref_dummy.upgrade() {
+            // no more references, let this task end
+            return;
+        }
+
+        match con_mgr.process() {
+            DidWork => tokio::task::yield_now().await,
+            NoWork => tokio::time::delay_for(std::time::Duration::from_millis(5)).await,
+            EndTask => return,
+        }
+    }
+}
+
+/// ConnectionMgr tracks a set of open websocket connections
+/// allowing you to send data to them and checking them for incoming data
+pub struct ConnectionMgr {
+    cmd_recv: CmdRecv,
+    evt_send_to_parent: EvtSend,
+    evt_send_from_children: EvtSend,
+    evt_recv_from_children: EvtRecv,
+    wss_map: std::collections::HashMap<Lib3hUri, CmdSend>,
+}
+
+impl ConnectionMgr {
+    /// spawn a new connection manager task, returning a handle for controlling it
+    /// and a receiving channel for any incoming data
+    pub fn new() -> (ConnectionMgrHandle, ConnectionMgrEventRecv) {
+        let (evt_p_send, evt_p_recv) = tokio::sync::mpsc::unbounded_channel();
+        let (evt_c_send, evt_c_recv) = tokio::sync::mpsc::unbounded_channel();
+        let (cmd_send, cmd_recv) = tokio::sync::mpsc::unbounded_channel();
+
+        let ref_dummy = Arc::new(());
+
+        let weak_ref_dummy = Arc::downgrade(&ref_dummy);
+
+        let con_mgr = ConnectionMgr {
+            cmd_recv,
+            evt_send_to_parent: evt_p_send,
+            evt_send_from_children: evt_c_send,
+            evt_recv_from_children: evt_c_recv,
+            wss_map: std::collections::HashMap::new(),
+        };
+
+        tokio::task::spawn(con_mgr_task(con_mgr, weak_ref_dummy));
+
+        (ConnectionMgrHandle::new(ref_dummy, cmd_send), evt_p_recv)
+    }
+
+    /// internal check our channels
+    fn process(&mut self) -> ConMgrResult {
```

Comment:
should we look into having process async as well?

Reply:
process is called from an async function - it performs one loop and doesn't need to await anything.

If we ever need to await something here it would be easy to add the async keyword :+1:

Reply:
nice!